### PR TITLE
Reuse refcount=0 lock in the LockPool

### LINKED ISSUE #13048: Reuse lock resource in lockPool to reduce high frequency new object
**Is your feature request related to a problem? Please describe.**

Now, alluxio LockPool is a great feature to reduce heap space while invent fine grain lock, but it can new lock resource and drop it in a high frequency.

**Describe the solution you'd like**
Use Resource pool to solve this issue.

**Describe alternatives you've considered**

Add a Resource pool in the lockPool, let's say it `resuePool`.

**Urgency**
Normal

**Additional context**
NO

### PR DESCRIPTION
**Pull request by maobaolong**
Fix #13048

### CODE REVIEW

File: core/common/src/main/java/alluxio/collections/LockPool.java [hunk 1]
```diff
@@ -256,7 +269,9 @@ private Resource getResource(K key) {
         v.mIsAccessed = true;
         return v;
       }
-      return new Resource(mDefaultLoader.apply(k));
+      Resource res = mResueLockPool.acquire();
```
Comment (apc999) [inline core/common/src/main/java/alluxio/collections/LockPool.java:272]: the behavior will be changed here, right? if we run out of the `mReuseLockPool`, will this `getResource` block?

Comment (maobaolong) [reply core/common/src/main/java/alluxio/collections/LockPool.java]: Yeah, it could be blocked

**Comment by maobaolong** [comment]
@apc999 Thank you for your review. 

I add a new api `acquireWithoutBlocking` to the `ResourcePool`, so that it can acquire object without blocking.

File: core/common/src/main/java/alluxio/resource/ResourcePool.java [hunk 2]
```diff
@@ -141,6 +141,22 @@ public T acquire(long time, TimeUnit unit) {
...
+   * @return a resource taken from the pool
+   */
+  public T acquireWithoutBlocking() {
+    // Try to take a resource without blocking
+    T resource = mResources.poll();
+    if (resource != null) {
+      return resource;
+    }
+    return createNewResource();
+  }
```
Comment (apc999) [inline core/common/src/main/java/alluxio/resource/ResourcePool.java:158]: Move the creation logic to caller side?
In this way, this `ResourcePool` provides a more isolated and independent function, just dealing with a pool of resources with different ways to acquire these resource (blocking or non-blocking)

```suggestion
  @Null
  public T acquireWithoutBlocking() {
    // Try to take a resource without blocking
    T resource = mResources.poll();
    if (resource != null) {
      return resource;
    }
    return null;
  }
```

File: core/common/src/main/java/alluxio/collections/LockPool.java [hunk 3]
```diff
@@ -58,7 +59,18 @@
   private static final String EVICTOR_THREAD_NAME = "LockPool Evictor";
 
   private final Map<K, Resource> mPool;
-  private final Function<? super K, ? extends ReentrantReadWriteLock> mDefaultLoader;
+  // TODO(baoloongmao): make it configurable before merge this PR.
```
Comment (apc999) [inline core/common/src/main/java/alluxio/collections/LockPool.java:62]: it's fine to hard code this constant for now.
unless you feel strong there is a need to configure this parameter in practise

File: core/common/src/main/java/alluxio/collections/LockPool.java [hunk 4]
```diff
@@ -256,7 +269,9 @@ private Resource getResource(K key) {
         v.mIsAccessed = true;
         return v;
       }
-      return new Resource(mDefaultLoader.apply(k));
+      Resource res = mResueLockPool.acquireWithoutBlocking();
```
Comment (apc999) [inline core/common/src/main/java/alluxio/collections/LockPool.java:272]: Once we change the `acquireWithoutBlocking` based on my suggestion, we should have it here:

```java
Resource res = mResueLockPool.acquireWithoutBlocking();
if (res == null) {
  res = mDefaultLoader.get();
}

File: core/common/src/main/java/alluxio/collections/LockPool.java (hunk 3 above)
Comment (maobaolong) [reply core/common/src/main/java/alluxio/collections/LockPool.java:62]: OK

File: core/common/src/main/java/alluxio/collections/LockPool.java (hunk 4 above)
Comment (maobaolong) [reply core/common/src/main/java/alluxio/collections/LockPool.java:272]: Done

**Review by apc999 — APPROVED** [review]

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Pull request merged**